src/read.c

```diff
@@ -2718,112 +2718,79 @@ void ReadEvalError(void)
 }
 
 
-/****************************************************************************
-**
-**  Reader state -- the next group of functions are used to "push" the
-**  current interpreter state allowing GAP code to be interpreted in the
-**  middle of other code. This is used, for instance, in the command-line
-**  editor.
-*/
-
-
-struct SavedReaderState {
-  UInt                userHasQuit;
-  syJmp_buf           readJmpError;
-  UInt                nrError;
-};
-
-static void SaveReaderState(struct SavedReaderState *s) {
-  s->userHasQuit = STATE(UserHasQuit);
-  s->nrError = STATE(NrError);
-  memcpy( s->readJmpError, STATE(ReadJmpError), sizeof(syJmp_buf) );
-}
-
-static void ClearReaderState(void ) {
-  STATE(UserHasQuit) = 0;
-  STATE(NrError) = 0;
-}
-
-static void RestoreReaderState(const struct SavedReaderState *s) {
-  memcpy( STATE(ReadJmpError), s->readJmpError, sizeof(syJmp_buf) );
-  STATE(UserHasQuit) = s->userHasQuit;
-  STATE(NrError) = s->nrError;
-}
-
-
 /****************************************************************************
 **
 *F  Call0ArgsInNewReader(Obj f)  . . . . . . . . . . . . call a GAP function
 **
 **  The current reader context is saved and a new one is started.
 */
 Obj Call0ArgsInNewReader(Obj f)
-
 {
-  struct SavedReaderState s;
-  IntrState intr = { 0, 0, 0, 0 };
-  Obj result;
+    syJmp_buf readJmpError;
+    volatile Obj result = 0;
 
-  /* remember the old reader context                                     */
-  SaveReaderState(&s);
+    // remember the old state
+    UInt userHasQuit = STATE(UserHasQuit);
+    UInt nrError = STATE(NrError);
+    Bag  oldLvars = STATE(CurrLVars);
+    memcpy(readJmpError, STATE(ReadJmpError), sizeof(syJmp_buf));
 
-  // initialize everything and begin an interpreter
-  ClearReaderState();
-  IntrBegin(&intr,  STATE(BottomLVars) );
+    // initialize everything
+    STATE(UserHasQuit) = 0;
+    STATE(NrError) = 0;
+    SWITCH_TO_OLD_LVARS(STATE(BottomLVars));
 
-  TRY_IF_NO_ERROR {
-    result = CALL_0ARGS(f);
-    PushVoidObj(&intr);
-    /* end the interpreter                                                 */
-    IntrEnd(&intr, 0, NULL);
-  }
-  CATCH_ERROR {
-    result = 0;
-    IntrEnd(&intr, 1, NULL);
+    TRY_IF_NO_ERROR
+    {
+        result = CALL_0ARGS(f);
+    }
+
+    // switch back to the old state
+    SWITCH_TO_OLD_LVARS(oldLvars);
     ClearError();
-  }
+    memcpy(STATE(ReadJmpError), readJmpError, sizeof(syJmp_buf));
+    STATE(UserHasQuit) = userHasQuit;
+    STATE(NrError) = nrError;
 
-  /* switch back to the old reader context                               */
-  RestoreReaderState(&s);
-  return result;
+    return result;
 }
 
+
 /****************************************************************************
 **
 *F  Call1ArgsInNewReader(Obj f,Obj a) . . . . . . . . . . call a GAP function
 **
 **  The current reader context is saved and a new one is started.
 */
-Obj Call1ArgsInNewReader(Obj f,Obj a)
-
+Obj Call1ArgsInNewReader(Obj f, Obj a)
```

Comment:
I don't like the complexity and similarity of Call0ArgsInNewReader and Call1ArgsInNewReader now SaveReaderState has been removed.

I'd prefer it either get reintroduced, or these two functions get a common implementation (which shouldn't be too hard, as they only differ on one line?)

Reply:
Yeah, this part of the PR definitely is unfinished. I might indeed revive `SaveReaderState`. (For some background, the single commit in this PR is the result of squashing several commits, minus some changes that were already merged otherwise, plus some extra tweaks. In the original series of commits, I simply did incremental changes as they became "obvious", the end result was not clear to me. As it is now, the minimal diff would just move the `CurrLVars` saving/restoring to `SaveReaderState`. Though I also wonder about the name of that function. And also whether perhaps we ought to save&restore more stuff, e.g. `STATE(Tilde)`. In the end, shouldn't `ReadEvalCommand` and the code here save&restore about the same things? Hmmm

Reply:
Updated